### PR TITLE
add cf.js data types

### PR DESCRIPTION
**Pull request by patience-tema-baron**
Moves the cf.js types into `src/types` and adds the new data types specified [here](https://github.com/counterfactual/monorepo/blob/master/packages/cf.js/API_REFERENCE.md#data-types).

### CODE REVIEW

File: packages/cf.js/src/types/structs.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+export interface AppABIEncodings {
+  stateEncoding: string;
+  actionEncoding?: string;
+}
+
+export interface BlockchainAsset {
+  assetType: number;
+  token?: string;
```
Comment (mvanderh) [inline packages/cf.js/src/types/structs.ts]: Make this an Address

File: packages/cf.js/src/types/simple-types.ts [hunk 2]
```diff
@@ -2,3 +2,6 @@ export type ABIEncoding = string;
 export type AppInstanceID = string;
 export type Address = string;
 export type Bytes32 = string;
+
+export type AppState = object;
+export type AppAction = object;
```
Comment (mvanderh) [inline packages/cf.js/src/types/simple-types.ts]: Technically these could also be simple `numbers` or `bytes` which wouldn't be translated into Typescript as `object`s. Better to make these `any`

File: packages/cf.js/src/types/structs.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  abiEncodings: AppABIEncodings;
+  asset: BlockchainAsset;
+  myDeposit: BigNumber;
+  peerDeposit: BigNumber;
+  timeout: BigNumber;
+}
+
+export interface AppABIEncodings {
+  stateEncoding: string;
+  actionEncoding?: string;
```
Comment (mvanderh) [inline packages/cf.js/src/types/structs.ts]: These should be `ABIEncoding`s

File: packages/cf.js/src/types/structs.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import { ABIEncoding, Address } from "./simple-types";
+import { BigNumber } from "ethers/utils";
+
+export interface AppDefinition {
```
Comment (mvanderh) [inline packages/cf.js/src/types/structs.ts]: Do we still need this?

File: packages/cf.js/src/types/structs.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  timeout: BigNumber;
+}
+
+export interface AppABIEncodings {
+  stateEncoding: string;
+  actionEncoding?: string;
+}
+
+export interface BlockchainAsset {
+  assetType: number;
```
Comment (mvanderh) [inline packages/cf.js/src/types/structs.ts]: Create an enum with the possible values to make this easier to read.

File: packages/cf.js/src/types/structs.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { ABIEncoding, Address } from "./simple-types";
+import { BigNumber } from "ethers/utils";
+
```
Comment (mvanderh) [inline packages/cf.js/src/types/structs.ts]: Add a link to the source documentation here

File: packages/cf.js/src/types/structs.ts (hunk 4 above)
Comment (cf19drofxots) [inline packages/cf.js/src/types/structs.ts]: Structs are typically more than just interfaces, so having interfaces in a structs file seems kinda weird.

Perhaps either just put them in the `src/types` file or in `src/types/index.ts`?

File: packages/cf.js/src/types/structs.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { ABIEncoding, Address } from "./simple-types";
```
Comment (mvanderh) [inline packages/cf.js/src/types/structs.ts]: Maybe rename this file to something more specific e.g. protocol-types?

File: packages/cf.js/src/types/structs.ts (hunk 4 above)
Comment (mvanderh) [reply packages/cf.js/src/types/structs.ts]: I don't like putting real code in index.ts. Prefer index.ts just be an exports aggregator.

Comment (cf19drofxots) [reply packages/cf.js/src/types/structs.ts]: Yeah me neither actually. Was a poor suggestion.

Comment (patience-tema-baron) [reply packages/cf.js/src/types/structs.ts]: We're still using it throughout cf.js, including sometimes referencing `appDefinition.address`. We couldn't remove it without breaking that code, and there's no obvious replacement for it either.

**Pull request merged**